File: src/main/java/org/nyxlang/interpreter/visitor/IfVisitor.java
package org.nyxlang.interpreter.visitor;

import org.nyxlang.interpreter.IInterpreter;
import org.nyxlang.interpreter.exception.VisitorException;
import org.nyxlang.parser.ast.INode;
import org.nyxlang.parser.ast.IfCase;
import org.nyxlang.parser.ast.IfNode;


/**
 * Interprets the conditional logic.
 */
public class IfVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return IfNode.class == node.getClass();
    }

    @Override
    public Object visit(IInterpreter interpreter, INode node) throws VisitorException {
        var ifNode = (IfNode) node;
        for (IfCase ifCase : ifNode.getCases()) {
            var conditionResult = interpreter.evalNode(ifCase.getCondition());
            if (Boolean.TRUE.equals(conditionResult)) {
                return interpreter.evalNode(ifCase.getBody());
            }
        }
        if (ifNode.getElseCase() != null) {
            return interpreter.evalNode(ifNode.getElseCase());
        }
        return null;
    }
}

File: src/main/java/org/nyxlang/interpreter/visitor/IfVisitor.kt
package org.nyxlang.interpreter.visitor

import org.nyxlang.interpreter.IInterpreter
import org.nyxlang.interpreter.result.IRuntimeResult
import org.nyxlang.interpreter.result.emptyResult
import org.nyxlang.parser.ast.ConditionalNode
import org.nyxlang.parser.ast.INode

/**
 * Interprets the conditional logic.
 */
class IfVisitor : IVisitor {
    override fun matches(node: INode) = ConditionalNode::class == node::class
    override fun visit(interpreter: IInterpreter, node: INode): IRuntimeResult {
        val ifNode = node as ConditionalNode
        for (ifCase in ifNode.cases) {
            val conditionResult = interpreter.evalNode(ifCase.condition).data
            if (conditionResult == true) {
                return interpreter.evalNode(ifCase.body)
            }
        }
        if (ifNode.elseCase != null) return interpreter.evalNode(ifNode.elseCase)
        return emptyResult()
    }
}